ntp-daemon/src/main.rs

```diff
@@ -1,169 +1,165 @@
-    ClockController, ClockUpdateResult, FilterAndCombine, NtpInstant, PollInterval, SystemConfig,
+    ClockController, ClockUpdateResult, FilterAndCombine, NtpInstant, PeerSnapshot, SystemConfig,
     SystemSnapshot,
 };
-use peer::{start_peer, PeerChannels};
+use peer::{start_peer, MsgForSystem, ResetEpoch};
 use tracing::info;
 
-use std::error::Error;
+use std::{error::Error, sync::Arc};
 use tokio::sync::watch;
 
+#[derive(Debug, Clone, Copy)]
+enum PeerState {
+    Demobilized,
+    AwaitingReset,
+    Valid(PeerSnapshot),
+}
+
+impl PeerState {
+    const fn get_snapshot(&self) -> Option<PeerSnapshot> {
+        match self {
+            PeerState::Demobilized | PeerState::AwaitingReset => None,
+            PeerState::Valid(snapshot) => Some(*snapshot),
+        }
+    }
+
+    fn reset(&mut self) {
+        *self = match self {
+            PeerState::Demobilized => PeerState::Demobilized,
+            PeerState::AwaitingReset => PeerState::AwaitingReset,
+            PeerState::Valid(_) => PeerState::AwaitingReset,
+        };
+    }
+}
+
 #[tokio::main]
 async fn main() -> Result<(), Box<dyn Error>> {
     tracing_subscriber::fmt::init();
 
     let config = SystemConfig::default();
 
-    // channel for sending updated system state to the peers
-    let (system_tx, system_rx) = watch::channel::<SystemSnapshot>(SystemSnapshot::default());
+    // shares the system state with all peers
+    let global_system_snapshot = Arc::new(tokio::sync::RwLock::new(SystemSnapshot::default()));
 
-    // channel to send the reset signal to all peers
-    let (reset_tx, reset_rx) = watch::channel::<u64>(0);
-    let mut last_reset_index: u64 = 0;
+    // send the reset signal to all peers
+    let mut reset_epoch: ResetEpoch = ResetEpoch::default();
+    let (reset_tx, reset_rx) = watch::channel::<ResetEpoch>(reset_epoch);
+
+    // receive peer snapshots from all peers
+    let (msg_for_system_tx, mut msg_for_system_rx) = tokio::sync::mpsc::channel::<MsgForSystem>(32);
 
     let mut controller = ClockController::new(UnixNtpClock::new());
 
-    let new_peer = |address| {
+    let peer_addresses = [
+        "0.pool.ntp.org:123",
+        "1.pool.ntp.org:123",
+        "2.pool.ntp.org:123",
+        "3.pool.ntp.org:123",
+    ];
+
+    for (index, address) in peer_addresses.iter().enumerate() {
         start_peer(
+            index,
             address,
             UnixNtpClock::new(),
             config,
-            system_rx.clone(),
+            msg_for_system_tx.clone(),
+            global_system_snapshot.clone(),
             reset_rx.clone(),
         )
-    };
+        .await
+        .unwrap();
+    }
 
-    let mut peers = vec![
-        new_peer("0.pool.ntp.org:123").await.unwrap(),
-        new_peer("1.pool.ntp.org:123").await.unwrap(),
-        new_peer("2.pool.ntp.org:123").await.unwrap(),
-        new_peer("3.pool.ntp.org:123").await.unwrap(),
-    ];
+    let mut peers = vec![PeerState::AwaitingReset; peer_addresses.len()];
```

Comment:
It feels to me like this would be cleaner more like this:

- Rename `PeerState` to `PeerStatus`
- Define `struct PeerState { status: PeerStatus, last: PeerSnapshot }`
- Define `struct Peers { peers: Vec<PeerState> }`

In general I would encourage making the system's state an explicit `struct`, which I think will make it clearer what state is involved and how that relates to each other.

Reply:
there is some stuff I like in here, but one thing I'm strongly against: making invalid states representable. What you are proposing with splitting the `PeerStatus` from the `last: PeerSnapshot` is effectively creating an untagged union. That's ... bad.

Reply:
I'm not sure what you mean, since in your current iteration you have `snapshots` and `peers` as separate `Vec`s, so my proposal doesn't change the design in that regard. (Unless I used the wrong type name for whatever thing `snapshots` contains.)

Reply:
Djc, I think folkert meant your suggestion for the struct peerstate, which seems to suggest separating the peersnapshot from the peerstatus. As a peersnapshot is really only useful for one specific status (the valid status) doing that would result in a datastructure that allows values to be created that are logically invalid, in effect the untagged union folkert is talking about.

Reply:
yes, exactly. Anyway I think the rest of the suggestions here have worked well and this is now outdated